url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/SIDMixer.java

```diff
@@ -0,0 +1,721 @@
+    recalcFilter();
+  }
+
+  public void setFilterCtrl(int data) {
+    lpOn = (data & 0x10) > 0;
+    bpOn = (data & 0x20) > 0;
+    hpOn = (data & 0x40) > 0;
+    if (DEBUG) {
+      System.out.println("Filter: " + data + " => " + (lpOn ? "LP " : "") +
+          (bpOn ? "BP " : "") + (hpOn ? "HP " : ""));
+    }
+    recalcFilter();
+  }
+
+  public void setFilterOn(int on) {
+    filterOn = on;
+  }
+
+  public void setMoogFilterOn(boolean on) {
+//  moog = on;
+  }
+
+  public boolean isMoogFilterOn() {
+    return false; // moog;
+  }
+
+  public void setMoogResonance(int percent) {
+//  moogResonance = percent;
+  }
+
+  public int getMoogResonance() {
+    return 0; //moogResonance;
+  }
+
+  public void setMoogCutoff(int frq) {
+//  moogCutoff = frq;
+  }
+
+  public int getMoogCutoff() {
+    return 0;// moogCutoff;
+  }
+
+  public void setMoogSpeed(int hz10) {
+//  mfrq = hz10;
+  }
+
+  public int getMoogSpeed() {
+    return 0;//mfrq;
+  }
+
+  public void setMoogDepth(int dep) {
+    //mdep = dep;
+  }
+
+  public int getMoogDepth() {
+    return 0; //mdep;
+  }
+
+  public void setVolume(int vol) {
+    sidVol = vol;
+  }
+
+  public void setVolume(int vol, long cycles) {
+    if (lastCycles > 0) {
+      int pos = (int) ((cycles - lastCycles) /
+          (SIDVoice6581.CYCLES_PER_SAMPLE + 10));
+      sidVolArr[pos % SIDVoice6581.GENLEN] = vol;
+    } else {
+      sidVol = vol;
+    }
+  }
+
+  private void recalcFilter() {
+    int fCutoff = 30 + (12000 * cutoff / 2048);
+    w0 = (long) (2 * Math.PI * fCutoff * 1.048576);
+    div1024Q = (int) (1024.0 / (0.707 + 1.0 * resonance / 15));
+  }
+
+
+  public void stop() {
+    setVolume(0);
+  }
+
+  public void reset() {
+    exVo = 0;
+    exVhp = 0;
+    exVlp = 0;
+
+    cutoff = 0;
+    resonance = 0;
+    w0 = 0;
+    div1024Q = 0;
+    filterOn = 0;
+
+    maxefx = 0;
+    minefx = 0;
+
+    for (int i = 0, n = SIDVoice6581.GENLEN; i < n; i++) {
+      sidVolArr[i] = -1;
+    }
+    setVolume(15);
+    recalcFilter();
+  }
+
+
+  public void printStatus() {
+    System.out.println("SIDMixer  ----------------------------");
+    System.out.println("Volume: " + sidVol);
+    System.out.println("FilterOn: " + filterOn);
+    System.out.println("Cutoff: " + cutoff);
+    System.out.println("Resonance: " + resonance);
+
+    System.out.println("Max Efx:" + maxefx);
+    System.out.println("Min Efx:" + minefx);
+//  moogFilter.printStatus();
+
+  }
+
+  public void setFullSpeed(boolean fs) {
+    System.out.println("Set full speed: " + fs);
+    fullSpeed = fs;
+  }
+
+  public boolean fullSpeed() {
+    return fullSpeed;
+  }
+
+  public int[] getBuffer() {
+    return intBuffer;
+  }
+
+
+
+  public static final int SLEEP_SYNC = 1;
+  private int sleep = 100;
+  private int syncMode = 0;//SLEEP_SYNC;
+  private double avg = 10;
+  private long lastTime = System.currentTimeMillis();
+  private long micros = 0;
+
+  public boolean updateSound(long cycles) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 36. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 540-575 into a separate method.